Connection fixes

vspinu opened the pull request:
A range of connection related fixes. One commit per issue. 

I will fix the last issue regarding auto-association to projects and jars in another day or two.

bbatsov commented on cider.el:
```diff
@@ -1208,6 +1215,29 @@ non-nil, don't start if ClojureScript requirements are not met."
+                            sessions)))
+
+    (when session
+      (unless (y-or-n-p "A session with the same parameters exists.  You can connect a sibling instead.  Proceed? ")
```
Maybe you can also print `session` as part of the message?

vspinu replied:
Could be, just a name though. Putting the full session is an overkill; too much stuff.

bbatsov replied:
👍

bbatsov commented on cider.el:
```diff
@@ -1208,6 +1215,29 @@ non-nil, don't start if ClojureScript requirements are not met."
                               cider-offer-to-open-cljs-app-in-browser)
                      (cider--offer-to-open-app-in-browser nrepl-server-buffer)))))))
 
+(defun cider--check-same-kind-session (params)
```
I'm not sure if `kind` is the right word for this. Maybe `check-for-existing` or `check-for-similar`?

vspinu replied:
`check-for-existing` sound right to me.

bbatsov replied:
👍

bbatsov commented on cider-connection.el:
```diff
@@ -639,6 +639,17 @@ function with the repl buffer set as current."
 
+(defun cider--no-repls-user-error (type)
+  "Throw \"No REPL\" user error customized for TYPE."
+  (let ((type (cond
```
Probably `pcase` would simplify this code. Or `cl-case`.

vspinu replied:
:) I don't see how it could be even simpler than that, it's a bare-bones conditional already.

bbatsov commented:
Overall the changes look good to me. Probably some simple test for the first and the last fix would be handy to make sure we don't accidentally break this down the road. 

I'm not sure I like the term `pending-cljs`, but I can't suggest something better of the top of my head. I'd like some term that suggests better that some REPL upgrade is in progress. I'm also wondering how are going to recover REPLs that for some reason don't managed to complete their "upgrade".

vspinu commented:
> I'm not sure I like the term pending-cljs,

Yeh. I actually did some digging on this one with synonyms et. al. But nothing better came up. I started with `uninitialized-cljs` but that's way too long for the modeline. 

> I'm also wondering how are going to recover REPLs that for some reason don't managed to complete their "upgrade".

We don't need to. If upgrade failed the repl returns to clj automatically on the first state message.

vspinu commented:
I have addressed the comments and added a test for the last commit. Writing tests for the first issue is quite pointless. One would need to overwrite every single function in `cider-connect-sibling-cljs`, thus tightly coupling the test with the implementation.

dpsutton commented on cider-connection.el:
```diff
@@ -92,6 +92,21 @@ PARAMS is a plist containing :host, :port, :server and other parameters for
+  "Gather all params for a SESSION."
+  (let (params)
+    (dolist (repl (cdr session))
+      (setq params (cider--gather-connect-params params repl)))
```
am i misreading this? This is a dolist and `setq` without accumulating previous values? Should this be `add-to-list` or `(setq params (cons params ...`?

dpsutton replied:
i missed `params` being passed into that function. ignore me :)
